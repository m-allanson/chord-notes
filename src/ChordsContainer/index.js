import React, { Component } from 'react'
import Chords from '../Chords'

const A = Symbol('A')
const A_SHARP = Symbol('A_SHARP')
const B = Symbol('B')
const C = Symbol('C')
const C_SHARP = Symbol('C_SHARP')
const D = Symbol('D')
const D_SHARP = Symbol('D_SHARP')
const E = Symbol('E')
const F = Symbol('F')
const F_SHARP = Symbol('F_SHARP')
const G = Symbol('G')
const G_SHARP = Symbol('G_SHARP')

const noteNames = new Map([
  [A, 'A'],
  [A_SHARP, 'A#'],
  [B, 'B'],
  [C, 'C'],
  [C_SHARP, 'C#'],
  [D, 'D'],
  [D_SHARP, 'D#'],
  [E, 'E'],
  [F, 'F'],
  [F_SHARP, 'F#'],
  [G, 'G'],
  [G_SHARP, 'G#'],
])

class ChordsContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {}
    this.setChordNotes = this.setChordNotes.bind(this)
    this.chordIntervals = new Map([
      ['major', [0, 4, 7]],
      ['minor', [0, 3, 7]],
      ['seventh', [0, 4, 7, 10]],
    ])
  }

  componentDidMount() {
    this.setChordNotes(A)
  }

  chordNotes(
    root,
    intervals,
    noteNames,
  ) {
    const notes = [...noteNames.keys()]
    const offset = notes.indexOf(root)
    const max = noteNames.size

    return intervals
      .map(index => {
        const offsetIndex = offset + index
        const normalIndex = (offsetIndex >= max)
          ? offsetIndex - max
          : offsetIndex
        return noteNames.get(notes[normalIndex])
      })
  }

  setChordNotes(root) {
    const {chordNotes, chordIntervals} = this
    const chords = [...chordIntervals]
      .map(([name, intervals]) => (
        [ name, chordNotes(root, intervals, noteNames) ]
      ))
    this.setState({
      chords: new Map(chords),
      root,
    })
    // , () => {console.log('state', this.state)}
  }

  render() {
    return (
      <div>
        <Chords
          noteNames={noteNames}
          onNoteClick={this.setChordNotes}
          chords={this.state.chords}
          root={this.state.root}
        />
      </div>
    )
  }
}

Chords.propTypes = {}

export default ChordsContainer
