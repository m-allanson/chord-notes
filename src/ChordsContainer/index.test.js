import React from 'react';
import ReactDOM from 'react-dom';
import ChordsContainer, { G } from './';

it('rotates an octave to have a root of G', () => {
  const component = new ChordsContainer()
  const {noteNames, getOctaveFromRoot} = component
  const root = G

  expect(
    getOctaveFromRoot(root, noteNames)
  ).toMatchSnapshot()
});

it('outputs the correct notes for G chords', () => {
  const component = new ChordsContainer()
  const {
    chordIntervals,
    noteNames,
    getOctaveFromRoot,
    getChords,
    fillFromFilter
  } = component
  const root = G
  const octave = getOctaveFromRoot(root, noteNames)

  expect(
    getChords(octave, chordIntervals, fillFromFilter)
  ).toMatchSnapshot()
});
