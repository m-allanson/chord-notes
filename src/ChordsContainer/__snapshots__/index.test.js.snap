exports[`test outputs the correct notes for G chords 1`] = `
Map {
  "major" => Array [
    Symbol(G),
    null,
    null,
    null,
    Symbol(B),
    null,
    null,
    Symbol(D),
    null,
    null,
    null,
    null,
  ],
  "minor" => Array [
    Symbol(G),
    null,
    null,
    Symbol(A_SHARP),
    null,
    null,
    null,
    Symbol(D),
    null,
    null,
    null,
    null,
  ],
  "seventh" => Array [
    Symbol(G),
    null,
    null,
    null,
    Symbol(B),
    null,
    null,
    Symbol(D),
    null,
    null,
    Symbol(F),
    null,
  ],
}
`;

exports[`test rotates an octave to have a root of G 1`] = `
Array [
  Symbol(G),
  Symbol(G_SHARP),
  Symbol(A),
  Symbol(A_SHARP),
  Symbol(B),
  Symbol(C),
  Symbol(C_SHARP),
  Symbol(D),
  Symbol(D_SHARP),
  Symbol(E),
  Symbol(F),
  Symbol(F_SHARP),
]
`;
